Sources/Algorithms/Chunked.swift

```diff
@@ -246,3 +246,133 @@ extension Collection {
     try chunked(on: projection, by: ==)
   }
 }
+
+//===----------------------------------------------------------------------===//
+// chunks(of:)
+//===----------------------------------------------------------------------===//
+
+/// A collection that presents the elements of its base collection
+/// in `SubSequence` chunks of any given size.
+///
+/// A ChunkedCollection is a lazy view on the base Collection, but it does not implicitly confer
+/// laziness on algorithms applied to its result.  In other words, for ordinary collections `c`:
+///
+/// * `c.chunks(of: 3)` does not create new storage
+/// * `c.chunks(of: 3).map(f)` maps eagerly and returns a new array
+/// * `c.lazy.chunks(of: 3).map(f)` maps lazily and returns a `LazyMapCollection`
+public struct ChunkedCollection<Base: Collection> {
+  
+  public typealias Element = Base.SubSequence
+  
+  public let base: Base
+  @usableFromInline
+  internal let _chunkCount: Int
+  
+  ///  Creates a view instance that presents the elements of `base`
+  ///  in `SubSequence` chunks of the given size.
+  ///
+  /// - Complexity: O(1)
+  @inlinable
+  internal init(_base: Base, _chunkCount: Int) {
+    self.base = _base
+    self._chunkCount = _chunkCount
+  }
+}
+
+extension ChunkedCollection: Collection {
+  public struct Index {
+    public let base: Base.Index
```

Comment:
This should also be internal.

There’s a bit of decision to make here between using this single index and using the range for the chunk. For a non-random-access collection, a single index makes using the subscript O(n), while a range moves that calculation to `startIndex`. Given that `startIndex` is so frequently touched for testing boundary conditions, the current implementation is probably the right one, but I don’t like that iterating over a collection this way requires traversing the base collection twice (once for each subscript access and once for each call to `index(after:)`).

Another approach would be to calculate `startIndex` at initialization and store it along with the chunk size. That way both `startIndex` and subscripting would always be O(1). What do you think?

Reply:
IIUC you're suggesting we store a Range in the index and calculate it in `init`, which would allow both `startIndex` and `subscript` to be O(1)? I like that idea!

Reply:
> Another approach would be to calculate `startIndex` at initialization and store it along with the chunk size. That way both `startIndex` and subscripting would always be O(1). What do you think?

I've experimented with this approach when putting up this PR, I like the idea ... but there are two points:
* Both start index and endIndex(lower bound) will have to be computed. 
* If we compute at initialization(start and end indexes) that the call then `[1,2,3].chunks(ofCount: 2)` becomes O(n) in non-`RandomAccessCollection` base.

Reply:
Just changed :)

Reply:
After implemented, I would say that I definitely like this approach!

Reply:
Great, thanks!

Reply:
@kylemacomber This is a more predictable version of what we've talked about for making `LazyFilter` better behaved — precalculating the start index would improve things there as well.

Reply:
I was thinking about this precomputation vs expensive `startIndex` vs expensive `self[startIndex]` trade-off, and I didn't know whether effectively precomputing `startIndex` was deemed "lazy enough" for lazy collections. I like this solution the most, to be clear 🙂 But at the same time, I've seen people (including myself) get confused by similar behaviour.

Reply:
I _think_ what I like about this approach is that it's more robust in the sense that we never end up with accidentally quadratic implementations.

Reply:
To me the best now is that iterations(which may be the most common operations in collections) are improved with this approach. Also it simplified the implementation which is always nice =]

Reply:
I've made all the adjustments so I think this is ready for review again now :))

Reply:
I’ve thought of caching `startIndex` for my filter-like collection wrappers before, but one part always bugged me enough to dissuade committing: what if the base collection is a reference type and it changed underneath me?!  The cached index may be invalid relative to the new base state; dereferencing may now even cause a crash.  I know 99.9% of the type the base collection will be a value type, and 99.9% of the time the base collection won’t be mutated while a chunk is active, so your solution of “so sad, too bad” is OK, but it’s still lousy from a purity perspective. The actual solution would be changing the language so general notifications are made when a reference object is mutated (so in this case, the cached `startIndex` can be recomputed), but that’s severe and I’m not sure it’s implementable.

Reply:
I think we have a point here... also I was thinking about this and this wouldn't be a big problem if we just compute the `startIndex` every time, as imagine most cases people base would likely to be an array or an lazy collection from a array source with `RandomAccessCollection` conditional conformance where base is random access so, in most cases `startIndex` wouldn't be expensive to compute. Also, I still like this range based approach because as Kyle mentioned is more robust and iterations are improved. 

> The actual solution would be changing the language so general notifications are made when a reference object is mutated (so in this case, the cached startIndex can be recomputed), but that’s severe and I’m not sure it’s implementable.

In C# we have a concept of `ObservableCollection` where mutations(well defined though an API) can be observed from anywhere. This not guarantee a notification of a custom collection mutation through a custom API, but I just remembered the concept, which is nice and very useful in the context of .NET framework =]

Reply:
For a reference-type collection or a collection of reference types, if there is a mutation between when a filter-like adapter is created and when it's traversed, I agree the result could be surprising and it's worth thinking carefully about our policy.

Note that for `chunks(ofCount:)`, mutations to reference type elements in a collection shouldn't change the behavior, and mutations to a reference type collection itself, possibly could be made to work. Though depending on our policy, we may decide we don't need to be defensive here.

Reply:
More concretely, my only concern with this approach was that certain lazy collections (most notably `LazyFilterCollection`) will run closures and evaluate elements when an index is advanced, and so precomputing the range of the first chunk can possibly run code that the user may not expect to be run at that stage. (Counterpoint: while lazy sequences are useful for short-circuiting, the user is very likely to at least evaluate the first element, in which case precomputing the start index doesn't have much of an impact.)

Calling `.prefix(n)` on a lazy collection has the same behavior, so we might have to conclude that such things are unavoidable/acceptable 🙂 In this case I think we should consider precomputing the first chunk of `chunked(on:)` and `chunked(by:)` as well.

Reply:
> Calling `.prefix(n)` on a lazy collection has the same behavior, so we might have to conclude that such things are unavoidable/acceptable 🙂 In this case I think we should consider precomputing the first chunk of `chunked(on:)` and `chunked(by:)` as well.

I agree, and I think at the end of the day as @kylemacomber mentioned it all depends if we decide to be really strict and  defensive or trade that of for performance :) 
Also, I still don't think that it would be a big problem to always compute the start index if we really want to be conservative. In either way let me know what you guys think and I'll be happy to make the changes if necessary =]

Reply:
> The actual solution would be changing the language so general notifications are made when a reference object is mutated (so in this case, the cached `startIndex` can be recomputed), but that’s severe and I’m not sure it’s implementable.

Another solution would be limiting the `Base` to be a value type. I think this was one of my inspirations to request adding an `AnyValue` poison protocol.